### PR TITLE
Issue 1

### PR DESCRIPTION
**Pull request by Alex-1985**
Temporary pull request to discuss ongoing chanes and share the feedback

### CODE REVIEW

File: src/main/java/jek/collections/IntArrayList.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @Override
+    public int size() {
+        return end;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (end == 0)
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java]: the stuff you put in braces is actually the boolean value and can be used as return values. So for your case the method can be simplified to 
return (end == 0);

File: src/main/java/jek/collections/IntArrayList.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    @Override
+    public Object[] toArray() {
+        return new Object[0];
+    }
+
+    @Override
+    public boolean add(Object o) {
+        if(end == ial.length) {                             //чекаем, если конец равен длинне
+            Object[] ial2 = new Object[ial.length*2];       //создаем новый массив и копируем
+            for (int i = 0; i < ial.length; i++)
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java]: Let's also start checking out what is there for us in the related libraries. Please check methods like System.arrayCopy
or Arrays.copy  - those methods are usually written in super efficient manner and so will make your code shorter(less lines) and faster to execute.

File: src/main/java/jek/collections/IntArrayList.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    @Override
+    public void add(int index, Object element) {
+
+    }
+
+    @Override
+    public Object remove(int index) {
+
+        if (index >= end || index < 0) {
+            throw new IndexOutOfBoundsException("ты пытаешься удалить то, чего у тебя нет!"); }
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java]: Good one! Seems like you understand how to read javadocs.

File: src/test/java/jek/collections/IntArrayListTest.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class IntArrayListTest {
+
+    @Test
+    void size() {
+        IntArrayList ial = new IntArrayList();
```
Comment (Alex-1985) [inline src/test/java/jek/collections/IntArrayListTest.java]: Good to start checking that your initial array is zero size

File: src/test/java/jek/collections/IntArrayListTest.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        ial.add(1);
+        ial.add(5);
+        ial.add(54);
+
+        assertEquals(3, ial.size());
+    }
+
+    @Test
+    void isEmpty() {
+        IntArrayList ial = new IntArrayList();
```
Comment (Alex-1985) [inline src/test/java/jek/collections/IntArrayListTest.java]: Good to start checking that your initial array is empty

File: src/test/java/jek/collections/IntArrayListTest.java [hunk 6]
```diff
@@ -107,10 +135,29 @@ void testRemove() {
...
+        ial.add(1);
+        ial.add(5);
+        ial.add(54);
+        ial.add(11);
+        ial.add(5);
+        ial.add(59);
+        ial.add(54);
+
+        assertEquals(4, ial.lastIndexOf(5));
     }
```
Comment (Alex-1985) [inline src/test/java/jek/collections/IntArrayListTest.java]: When you are storing or retrieving the object always think how do you want your application to behave when you are passing the null value. Test it also please. 
What should there be if there's no such element in your list.

File: src/test/java/jek/collections/IntArrayListTest.java [hunk 7]
```diff
@@ -7,7 +7,14 @@
 class IntArrayListTest {
 
     @Test
-    void size() {
+    void sizeShouldBeZero() {
+        IntArrayList ial = new IntArrayList();
```
Comment (Alex-1985) [inline src/test/java/jek/collections/IntArrayListTest.java]: Each of the methods begin with the same line. Please read about the @Setup, @Before and other related annotations and use them to write method that will initialize the arraylist beforehand.

File: src/main/java/jek/collections/IntArrayList.java [hunk 8]
```diff
@@ -100,22 +94,37 @@ public void add(int index, Object element) {
...
 
-        if (index >= end || index < 0) {
-            throw new IndexOutOfBoundsException("ты пытаешься удалить то, чего у тебя нет!"); }
+        if (index >= end || index < 0)
+            throw new IndexOutOfBoundsException("ты пытаешься удалить то, чего у тебя нет!");
 
+        Object removed = ial[index];
 
-     return null;
+        if (index < end-1)
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java:150]: Very nice thinking about necessity to decrease size of array? Why don't you just leave it as is... 


performance vise it's very poor. 
Copying array is very heavy weight operation. Also think how many times you increase your array during adding, and make it symmetrical for removal.

File: src/main/java/jek/collections/IntArrayList.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    /* вот запилил итератор, но еще не особо прочекал его, бо уже иду спать(
+    * я еще перепроверю, и то шо ты в комментах написал тож поправлю.
+    * но наверн уже на выхах, бо в чтв и птнц дохуя работаю.
+    * пис.
+    *
+    *
+    *  */
+
+    @Override
+    public Iterator iterator() {
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java:38]: While working on Iterators please pay attention to it's documentation from that perspective on how it should behave when somebody already got the iterator and then the list was changed.

File: src/main/java/jek/collections/IntArrayList.java [hunk 10]
```diff
@@ -20,6 +20,7 @@ public int size() {
 
     @Override
     public boolean isEmpty() {
+
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java:23]: keep an eye on what you commit. If you use Idea plugin should be easy to recheck all the changes before committing them.  Please also start making more senseful commit messages. Some changes is not informative. It's better to write something like "Working on review comments. Fixed a)... b)... c)...." or "Implemented method xxx with test"

File: src/main/java/jek/collections/IntArrayList.java [hunk 11]
```diff
@@ -293,4 +328,20 @@ public boolean containsAll(Collection c) {
...
+    private void increaseArray() {
+        if(end == ial.length) {
+            Object[] ial2 = new Object[ial.length*2];
+            System.arraycopy(ial, 0, ial2, 0, ial.length);
+            ial = ial2;
+        }
+    }
+
+    private void decreaseArray() {
+        if (end < (ial.length/2-1) && end > 10){
```
Comment (Alex-1985) [inline src/main/java/jek/collections/IntArrayList.java]: Why 10?

Comment (sergeyLiquid) [reply src/main/java/jek/collections/IntArrayList.java]: 10 was our initial size, so I decided not to step over it

File: src/main/java/jek/collections/IntArrayList.java (hunk 10 above)
Comment (sergeyLiquid) [reply src/main/java/jek/collections/IntArrayList.java:23]: committed with git only. next time will try with Idea plugin.

File: src/main/java/jek/collections/IntArrayList.java (hunk 11 above)
Comment (sergeyLiquid) [reply src/main/java/jek/collections/IntArrayList.java]: changed to 9, if you meant this

Comment (Alex-1985) [reply src/main/java/jek/collections/IntArrayList.java]: not really... the 10(or 9 in this case already)  is what called "magic number". That's one of types of problems which are identified during the code review. Usually 1 might be there if we're expecting only single entrance. 2 is also ok as you used to increase twice the size of initial array. Usually we move the magic numbers to constants. e.g. INITIAL_ARRAY_SIZE or MULTIPLY_FACTOR or whatever would be suitable.

**Comment by Alex-1985** [comment]
Commit comments are better now, thanks.

**Pull request merged**